### DISCUSSION TITLE
Lesson 12: testGetTokenAmountFromUsd (confusing usdAmount declaration)

### DISCUSSION THREAD

**Discussion by ChocLabEdi** [Cyfrin/foundry-full-course-cu#1159] [Q&A] [answered]
I have a question about the unit referred by Patrick in the tutorial in testGetTokenAmountFromUsd. Why would we set the testing variable as this (uint256 usdAmount = 100 ether;) instead of a value in usd? I thought the parameter for the function to be tested getTokenAmountFromUsd expects a value in usd instead of ether? I see how in the assertEq(amountWeth, expectedWeth) the appearance of ether occurs on both sides of the equation so they work out, but it looks a bit 'strange' to have '100 ether' as an usd amount, or am I missing something?

Or the test could be re-written as follows by removing ether in both lines? 

proposed updates to two lines of test codes:
```
        uint256 usdAmount = 100;  //which means 100 usd
         uint256 expectedWeth = 0.05;  //which means 0.05 eth
```

the test:
```
    function testGetTokenAmountFromUsd() public {
        // If we want $100 of WETH @ $2000/WETH, that would be 0.05 WETH
        uint256 usdAmount = 100 ether;
         uint256 expectedWeth = 0.05 ether;
        uint256 amountWeth = dsce.getTokenAmountFromUsd(weth, usdAmount);
        assertEq(amountWeth, expectedWeth);
    }

```

the function to be tested: 
```
    function getTokenAmountFromUsd(address token, uint256 usdAmount) public view returns (uint256) {
        AggregatorV3Interface priceFeed = AggregatorV3Interface(s_priceFeeds[token]);
        (, int256 price,,,) = priceFeed.latestRoundData();
        // $100e18 USD Debt
        // 1 ETH = 2000 USD
        // 1 ETH = 1e18 Wei
        // The returned value from Chainlink will be 2000 * 1e8
        // Most USD pairs have 8 decimals, so we will just pretend they all do
        return ((usdAmount * 1e18) / (uint256(price) * 1e10)); 
    }
```

**Comment by Niferu** [comment, marked as answer]
Hi @ChocLabEdi 

You must know that in solidity `100 ether` equals `100 ** 10^18`. Then in our function we have `expectedWeth`, which is amount of tokens received for given amount of usd paid. We are also setting ETH price to 2000 USD per 1 ETH(from our mock aggregator), which gives us following math equation: usdAmount/2000(1 ETH price in usd) -> this will give us amount of weth tokens (eth = weth) and exactly that we are doing in `testGetTokenAmountFromUsd` function. We are answering question how much ETH we will get for 100 000 000 000  000 000 000 USD. If you delete `ether` from tested function you will see syntax error due to solidity restriction of using decimals. That is why we are adding `ether` here to increase number that is real purpose as we are just checking if function calculate things properly. 

If you dont want to use `ether` just do below:

```solidity
function testGetTokenAmountFromUsd() public {
        uint256 usdAmount = 10000;
        // 10000 USD/ 2000 USD = 5 ETH
        uint256 expectedWeth = 5;
        uint256 actualWeth = nfre.getTokenAmountFromUsd(weth, usdAmount);

        assertEq(expectedWeth, actualWeth);
    }
```

**Reply by ChocLabEdi** [reply]
Got it. So the 'ether' in the original codes was merely used as a precision factor to get around the decimal places - that's where the confusion comes from. The way you suggested works. Or just create a separate precision factor (such as 1e6) more explicitly instead of 'ether' should help in clarity conceptually

**Reply by Niferu** [reply]
Yup exactly, glad I could clarify it for you :)
